core/trino-main/src/main/java/io/trino/operator/ExchangeOperator.java

```diff
@@ -197,6 +215,290 @@ public Page getOutput()
+                ExchangeManagerRegistry exchangeManagerRegistry)
+        {
+            this.directExchangeClientSupplier = requireNonNull(directExchangeClientSupplier, "directExchangeClientSupplier is null");
+            this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+            this.taskFailureListener = requireNonNull(taskFailureListener, "taskFailureListener is null");
+            this.retryPolicy = requireNonNull(retryPolicy, "retryPolicy is null");
+            this.exchangeManagerRegistry = requireNonNull(exchangeManagerRegistry, "exchangeManagerRegistry is null");
+        }
+
+        @Override
+        public SerializedPage pollPage()
+        {
+            if (closed.get()) {
+                return null;
+            }
+
+            ExchangeDataSource dataSource = delegate.get();
+            if (dataSource == null) {
+                return null;
+            }
+            return dataSource.pollPage();
+        }
+
+        @Override
+        public boolean isFinished()
+        {
+            if (closed.get()) {
+                return true;
+            }
+            ExchangeDataSource dataSource = delegate.get();
+            if (dataSource == null) {
+                return false;
+            }
+            return dataSource.isFinished();
+        }
+
+        @Override
+        public ListenableFuture<Void> isBlocked()
+        {
+            if (closed.get()) {
+                return immediateVoidFuture();
+            }
+            if (!initializationFuture.isDone()) {
+                return initializationFuture;
+            }
+            ExchangeDataSource dataSource = delegate.get();
+            if (dataSource == null) {
+                return immediateVoidFuture();
+            }
+            return dataSource.isBlocked();
+        }
+
+        @Override
+        public void addSplit(RemoteSplit remoteSplit)
+        {
+            SettableFuture<Void> future = null;
+            synchronized (this) {
+                if (closed.get()) {
+                    return;
+                }
+                ExchangeDataSource dataSource = delegate.get();
+                if (dataSource == null) {
+                    if (remoteSplit.getExchangeInput() instanceof DirectExchangeInput) {
+                        DirectExchangeClient directExchangeClient = directExchangeClientSupplier.get(systemMemoryContext, taskFailureListener, retryPolicy);
+                        dataSource = new DirectExchangeDataSource(directExchangeClient);
+                        dataSource.addSplit(remoteSplit);
+                    }
+                    else if (remoteSplit.getExchangeInput() instanceof ExternalExchangeInput) {
+                        ExternalExchangeInput externalExchangeInput = (ExternalExchangeInput) remoteSplit.getExchangeInput();
+                        ExchangeManager exchangeManager = exchangeManagerRegistry.getExchangeManager();
```

Comment:
If we statically can tell if we are operating in direct or external mode we should obtain instance of `ExchangeManager exchangeManager` when factory is being constructed.

Reply:
I assume it is related to this: https://github.com/trinodb/trino/pull/9818#discussion_r753088796